nscq/src/nscq_handler.rs

```diff
@@ -0,0 +1,285 @@
+            ptr::null_mut(),
+            0,
+        );
+        let rc = *LAST_RETURN_CODE.lock().unwrap();
+        if rc == 0 {
+            Ok(SHARED_TNVL_STATUS.lock().unwrap().clone())
+        } else {
+            Err(rc)
+        }
+    }
+
+    pub fn is_switch_tnvl_mode(&self, device: &str) -> Result<bool, NscqRc> {
+        let tnvl_status = self.get_switch_tnvl_status(device)?;
+        Ok(((tnvl_status >> TNVL_BIT_POSITION) & 1) == 1)
+    }
+
+    pub fn is_switch_lock_mode(&self, device: &str) -> Result<bool, NscqRc> {
+        let tnvl_status = self.get_switch_tnvl_status(device)?;
+        Ok(((tnvl_status >> LOCK_BIT_POSITION) & 1) == 1)
+    }
+
+    pub fn get_switch_attestation_certificate_chain(
+        &self,
+        device: &str,
+    ) -> Result<Vec<u8>, NscqRc> {
+        self.session.path_observe(
+            &format!("/{device}{ATTESTATION_CERTIFICATE_CHAIN_PATH}"),
+            NscqCallback::AttestationCertificate(_device_attestation_certificate_callback),
+            ptr::null_mut(),
+            0,
+        );
+        let rc = *LAST_RETURN_CODE.lock().unwrap();
+        if rc == 0 {
+            let shared_certificate_chain = SHARED_CERTIFICATE_CHAIN.lock().unwrap();
+            let certificate_chain = shared_certificate_chain
+                .get(device)
+                .map(|chain| chain.as_slice())
+                .unwrap_or(&[]);
+            Ok(certificate_chain.to_vec())
+        } else {
+            Err(rc)
+        }
+    }
+
+    pub fn get_all_switch_attestation_certificate_chain(
+        &self,
+    ) -> Result<HashMap<String, Vec<u8>>, NscqRc> {
+        self.session.path_observe(
+            &format!("{ALL_SWITCHES}{ATTESTATION_CERTIFICATE_CHAIN_PATH}"),
+            NscqCallback::AttestationCertificate(_device_attestation_certificate_callback),
+            ptr::null_mut(),
+            0,
+        );
+        let rc = *LAST_RETURN_CODE.lock().unwrap();
+        if rc == 0 {
+            Ok(SHARED_CERTIFICATE_CHAIN.lock().unwrap().clone())
+        } else {
+            Err(rc)
+        }
+    }
+
+    pub fn get_switch_attestation_report(
+        &self,
+        device: &str,
+        nonce: &[u8; 32],
+    ) -> Result<[u8; NSCQ_ATTESTATION_REPORT_SIZE], NscqRc> {
+        let nonce_ptr = nonce.as_ptr() as *mut c_void;
+        let nonce_len = nonce.len() as u32;
+        let input_arg = unsafe { &mut *nonce_ptr };
+        self.session.set_input(input_arg, nonce_len, 0);
+        self.session.path_observe(
+            &format!("/{device}{ATTESTATION_REPORT_PATH}"),
+            NscqCallback::AttestationReport(_device_attestation_report_callback),
+            ptr::null_mut(),
+            0,
+        );
+        let rc = *LAST_RETURN_CODE.lock().unwrap();
+        if rc == 0 {
+            let shared_attestation_report = SHARED_ATTESTATION_REPORT.lock().unwrap();
+            let report = shared_attestation_report
+                .get(device)
```

Comment:
When was the `shared_attestation_report` updated ?

Reply:
It was in the callback, now it's done differently.